doc/Frontend/syntax.org

```diff
@@ -0,0 +1,126 @@
+* Syntax
+#+BEGIN_SRC bnf
+  top-level ::= <type>
+              | <module-open>
+              | <type-class>
+              | <type-class-instance>
+              ; These four last, due to lack of words before
+              | <module-signature>
+              | <module>
+              | <signature>
+              | <function>
+
+  ;;; Functions ======================================================
+
+  function ::= <non-capital-name> <name>+ = <expression>
+
+  ; if we only have usage in arrows, we should allow a way to have
+  ; usage information of variables as well!
+  signature ::= <non-capital-name> <type-signature>
+              | <non-capital-name> <usage> <type-signature>
+
+  ;;; Types ==========================================================
+
+  type ::= <data-declaration>
+         | <type-declaration>
+         | <alias-declaration>
+
+  ; Should we let there be a refinement in adt type declaration?
+  data-declaration  ::= type <non-capital-name> : <arrow-type> = <adt>
+                      | type <non-capital-name>                = <adt>
+
+  type-declaration  ::= type <non-capital-name> = <name> <type-refine>
+  alias-declaration ::= type <non-capital-name> = <type-refine>
+
+  adt ::= <sum>*
```

Comment:
Hmm this will look like inductive families, at least in the top-level frontend, not like Haskell ADTs - see the bottom of p.17 in the language reference.

Reply:
These are alias declarations, not data declarations see,

```ebnf
  data-declaration  ::= type <non-capital-name> : <arrow-type> = <adt>
                      | type <non-capital-name>                = <adt>


  product ::= { <name-type-semi>+ <name-type>* }
            | { <name-type-semi>* }
            | <arrow-type>


  arrow-type ::= <type-refine>
               | <name> : <type-refine> <arrow> <arrow-type>
               | <type-refine>          <arrow> <arrow-type>
               | ( <arrow-type> )
```


namely the <arrow-type> part.

Though I should add

```ebnf
  product ::= { <name-type-semi>+ <name-type>* } <name-type>
            | { <name-type-semi>* } <name-type>
            | { <name-type-semi>+ <name-type>* }
            | { <name-type-semi>* }
            | <arrow-type>
```

This amounts to inductive family syntax.